### PR TITLE
a2gameio: Add support for the Softape Bright Pen

### PR DESCRIPTION
**Pull request by as-tb-dev**
Adds support for the Softape Bright Pen for the Apple ][ and ][+

Can be tested with these Softape games:
- Bright Pen Craps
- Roulette

### CODE REVIEW

File: src/devices/bus/a2gameio/brightpen.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+				// where the PPU is drawing on the CRT,
+				if (j <= vpos && j > vpos - sustain && (j != vpos || i <= hpos))
+					sum += pix.r() + pix.g() + pix.b();
+				scanned++;
+			}
+		}
+	}
+
+	// light detected if average brightness is above threshold
+	return (sum >= bright * scanned) && BIT(m_brightpen_point->read(), 0);
```
[the remark this answers is not in the document]
Comment (as-tb-dev) [reply src/devices/bus/a2gameio/brightpen.cpp:117]: Correct, it was.  Due to that, I had also copied the copyright-holders line at the top of this file and put a reference to the original NES file just ahead of this function call.  Is this method of attribution correct?  If not, I will adjust it.

[the remark this answers is not in the document]
Comment (cuavas) [reply src/devices/bus/a2gameio/brightpen.cpp:119]: Copying the copyright holders for a derivative work is the correct thing to do, but there are other issues.

Firstly, the comments aren’t really valid in this application.  The Apple II doesn’t have a “PPU”, for example.  That’s Nintendo’s terminology for sprite/tile rendering chips.

Secondly, there are significant differences between a lightpen and the photosensor gun used with the NES.
* A lightpen is held a lot closer to the screen and can sense whether it’s over an illuminated pixel far more accurately.
* A lightpen usually has a lower threshold for detecting a lit pixel than the NES photosensor gun.

I’d expect this to require more adjustment.

File: src/devices/bus/a2gameio/brightpen.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+{
+}
+
+// light detection logic based on zapper_sensor.cpp nes_zapper_sensor_device::detect_light
+
+int apple2_brightpen_device::sw0_r() {
+        int x = m_brightpen_x->read();
+        int y = m_brightpen_y->read();
+	int vpos = m_screen->vpos();
+	int hpos = m_screen->hpos();
```
[the remark this answers is not in the document]
Comment (rb6502) [reply src/devices/bus/a2gameio/brightpen.cpp]: Running MAME's "srcclean" tool on the source files is always a good idea prior to a submission.  It helps catch stuff like this before you see it on Github.

**Comment by rb6502** [comment]
I agree with Vas that this shouldn't need that big of a pixel radius (lightpens weren't the world's most precise things, but they were definitely better than the NES Zapper), and the comments about the NES PPU certainly don't apply.  Would you mind fiddling with it and seeing if a smaller radius and lower threshold works, and also correcting the comment?

File: src/devices/bus/a2gameio/brightpen.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (as-tb-dev) [reply src/devices/bus/a2gameio/brightpen.cpp]: Thanks.  I have built the tools recently as well and have started running it prior to check-ins.

**Comment by as-tb-dev** [comment]
I made improvements to the comment, reduced the radius and lowered the brightness threshold.  With these lower numbers, it still works with Roulette and Craps.

**Pull request merged**